### PR TITLE
CIFAR 10 ExecuTorch model fine-tuning on the Android app

### PR DESCRIPTION
**Pull request by IshanAryendu**
In this tutorial, we are trying to fine-tuning a CIFAR 10 model on an Android device using ExecuTorch PTE + PTD combo. This example is an extension of the the CIFAR training example from the official [GitHub repository](https://github.com/pytorch/executorch/tree/main/extension/training/examples/CIFAR).

### CODE REVIEW

File: cifar/README.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
+# ExecuTorch Model Fine-Tuning on Android App
+
+In this tutorial, we will be fine-tuning a CIFAR 10 model on an android app using ExecuTorch.
+
+## Environment Setup
+
+For the android environment setup, follow these steps:
+
+1. Install Open JDK using: `brew install openjdk@17`
```
Comment (jathu) [inline cifar/README.md:9]: Not everyone will be using macOS. Let's just delegate the instruction to [OpenJDK](https://openjdk.org/). However, I also think we don't really need to do this — it might be safe to assume Android engineers already know about this (+ the other instructions related to Android in general below)

File: cifar/README.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+2. Download and install the latest version of [Android Studio](https://developer.android.com/studio/)  .
+3. Start Android Studio and open the **Settings** dialog (gear icon on the bottom left).
+4. Navigate to **Languages & Frameworks**, then **Android SDK**.
+5. In the **SDK Platforms** tab, check **Android 14.0 (“UpsideDownCake”)** and **Android API 36**.
+6. In the **SDK Tools** tab, check **Android SDK Build-Tools 36**, **NDK (Side by side)**, **Android SDK Command-line Tools (latest)**, **CMake**, **Android Emulator**, and **Android SDK Platform-Tools**.
+7. Export the paths for these to your environment or add them to the `.bashrc` or `.zshrc` files:
+```bash
+# JAVA config
+export JAVA_HOME=/opt/homebrew/opt/openjdk@17/
```
Comment (jathu) [inline cifar/README.md:18]: This is macOS and version specific, we shouldn't include that. We can reduce these steps to:

```
Ensure the following environment variables are set: `JAVA_HOME`, `ANDROID_NDK`, `ANDROID_SDK`
```

Better yet we can enforce that in our build scripts ;)

File: cifar/README.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+# Cmake config
+export PATH=$ANDROID_HOME/cmake/4.0.2/bin:$PATH
+```
+
+**NOTE** For the updated steps for building the dependencies refer to the official repository over [here](https://github.com/pytorch/executorch/blob/main/extension/android/README.md).
+
+To install ExecuTorch in a python environment we can use the following commands in a new terminal:
+
+```bash
+$ git clone https://github.com/pytorch/executorch.git
```
Comment (jathu) [inline cifar/README.md:38]:
```suggestion
$ git clone https://github.com/pytorch/executorch.git --recurse-submodules
```

File: cifar/README.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+$ cd executorch
+$ uv venv --seed --prompt et --python 3.10
```
Comment (jathu) [inline cifar/README.md:40]: Not everyone has [uv](https://github.com/astral-sh/uv). We should suggest they download and use UV, or some other py env tool

File: cifar/README.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+$ source .venv/bin/activate
+$ which python
```
Comment (jathu) [inline cifar/README.md:42]: Not needed

File: cifar/README.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+$ git fetch origin
+$ git submodule sync --recursive
+$ git submodule update --init --recursive
+$ ./install_requirements.sh
```
Comment (jathu) [inline cifar/README.md:46]: We can remove this, `install_executorch.sh` calls this under the hood anyways

File: cifar/README.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+$ ./install_executorch.sh
+$ sh ./scripts/build_android_library.sh
+$ echo $ANDROID_HOME
```
Comment (jathu) [inline cifar/README.md:49]: Not needed

File: cifar/README.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+$ cd executorch
+$ uv venv --seed --prompt et --python 3.10
+$ source .venv/bin/activate
+$ which python
+$ git fetch origin
+$ git submodule sync --recursive
+$ git submodule update --init --recursive
+$ ./install_requirements.sh
+$ ./install_executorch.sh
+$ sh ./scripts/build_android_library.sh
```
Comment (jathu) [inline cifar/README.md:48]:
```suggestion
$ ./scripts/build_android_library.sh
```

File: cifar/README.md (hunk 4 above)
Comment (larryliu0820) [reply cifar/README.md:40]: Oh I didn't know `uv` works on ExecuTorch

File: cifar/app/src/main/java/com/example/democifar10/MainActivity.kt [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+        return bitmap
+    }
+
+    // Function to save a bitmap to a PNG file
+    private fun saveBitmapToPng(bitmap: Bitmap, file: File) {
+        try {
+            FileOutputStream(file).use { out ->
+                bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
+            }
+            Log.d("ExecuTorchApp", "Saved image to ${file.absolutePath}")
```
Comment (larryliu0820) [inline cifar/app/src/main/java/com/example/democifar10/MainActivity.kt:91]: Can you extract this tag out as a variable? If we need to change it in the future it would be much easier.

**Review by larryliu0820 — COMMENTED** [review]
@psiddh can you also take a look?

File: cifar/README.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+Finally, the `.aar` file can be found here:
+
+```bash
+<PARENT_DIRECTORY>/executorch/extension/android/executorch_android/build/outputs/aar/executorch_android-debug.aar
+```
+
+**Note:** We will rename this file to `executorch.aar` and copy it into the `libs` directory of the android app.
+
+## Creation of Android App
```
Comment (psiddh) [inline cifar/README.md:77]: Do we really need this step ? The current app layout has all the necessary files & structure.  Can we remove this section ? that way it can be less verbose

File: cifar/README.md (hunk 1 above)
Comment (IshanAryendu) [reply cifar/README.md]: I'll update this to direct the people to [OpenJDK](https://openjdk.org/) and follow the steps over there to configure their environment

File: cifar/app/src/main/java/com/example/democifar10/MainActivity.kt (hunk 9 above)
Comment (IshanAryendu) [reply cifar/app/src/main/java/com/example/democifar10/MainActivity.kt]: The variable debugTag stores the required tag for easier maintenance

File: cifar/README.md (hunk 10 above)
Comment (IshanAryendu) [reply cifar/README.md]: Made the recommended changes to the README file

File: cifar/android/CifarETTrainingDemo/README.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+If you run into errors for sdk path, complete the above steps in the [Trouble Shooting](#trouble-shooting) section before proceeding with the following steps:
+
+```bash
+$ sh ./scripts/build_android_library.sh
+$ sh ./extension/android/executorch_android/android_test_setup.sh
+$ ls ./extension/android/executorch_android/
+$ cd extension/android
+$ ./gradlew :executorch_android:testDebugUnitTest
+$ ./gradlew :executorch_android:connectedAndroidTest
+```
```
Comment (jathu) [inline cifar/android/CifarETTrainingDemo/README.md:51]: Do we really need to do this?

Comment (IshanAryendu) [reply cifar/android/CifarETTrainingDemo/README.md:54]: I've retained these steps as a sanity check

File: cifar/android/CifarETTrainingDemo/app/src/main/java/com/example/democifar10/ui/theme/Theme.kt [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+
+    /* Other default colors to override
+    background = Color(0xFFFFFBFE),
+    surface = Color(0xFFFFFBFE),
+    onPrimary = Color.White,
+    onSecondary = Color.White,
+    onTertiary = Color.White,
+    onBackground = Color(0xFF1C1B1F),
+    onSurface = Color(0xFF1C1B1F),
+    */
```
[the remark this answers is not in the document]
Comment (IshanAryendu) [reply cifar/android/CifarETTrainingDemo/app/src/main/java/com/example/democifar10/ui/theme/Theme.kt]: Removed these comments

File: cifar/android/CifarETTrainingDemo/README.md (hunk 11 above)
Comment (jathu) [reply cifar/android/CifarETTrainingDemo/README.md:54]: Sanity check for who/what?

Comment (IshanAryendu) [reply cifar/android/CifarETTrainingDemo/README.md:54]: Oh, these are used to run the unit tests and instrumented tests for the newly built modules

File: cifar/android/CifarETTrainingDemo/app/src/main/res/values/colors.xml [hunk 13]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (jathu) [inline cifar/android/CifarETTrainingDemo/app/src/main/res/values/colors.xml:1]: Do we need this?

File: cifar/android/CifarETTrainingDemo/README.md (hunk 11 above)
Comment (jathu) [reply cifar/android/CifarETTrainingDemo/README.md:54]: We should assume that the tests already pass — we don't run tests for libraries we use. If a user skips these tests, will the training example work? If so, let's remove it

File: cifar/android/CifarETTrainingDemo/app/src/main/res/values/colors.xml (hunk 13 above)
Comment (larryliu0820) [reply cifar/android/CifarETTrainingDemo/app/src/main/res/values/colors.xml:1]: oh maybe it's optional, since string.xml doesn't have it?

File: cifar/android/CifarETTrainingDemo/README.md [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+# Build the data and models
+$ mkdir -p executorch-examples/cifar/android/CifarETTrainingDemo/app/src/main/assets/
+$ python ./extension/training/examples/CIFAR/main.py --data-dir executorch-examples/cifar/android/CifarETTrainingDemo/app/src/main/assets/ --model-path executorch-examples/cifar/android/CifarETTrainingDemo/app/src/main/assets/cifar10_model.pth --pte-model-path executorch-examples/cifar/android/CifarETTrainingDemo/app/src/main/assets/cifar10_model.pte --split-pte-model-path executorch-examples/cifar/android/CifarETTrainingDemo/app/src/main/assets/cifar10_model_pte_only.pte --save-pt-json executorch-examples/cifar/android/CifarETTrainingDemo/app/src/main/assets/cifar10_pt.json --save-et-json executorch-examples/cifar/android/CifarETTrainingDemo/app/src/main/assets/cifar10_et.json --ptd-model-dir executorch-examples/cifar/android/CifarETTrainingDemo/app/src/main/assets/ --epochs 5 --fine-tune-epochs 10
+```
+
+If you run into errors for sdk path, complete the above steps in the [Trouble Shooting](#trouble-shooting) section before proceeding.
+
+## Creation of Android App
+
+1. Start with a clone of this repository and open the project in Android Studio.
```
Comment (psiddh) [inline cifar/android/CifarETTrainingDemo/README.md:63]: Much cleaner now, can add path to gradle file ?

File: cifar/android/CifarETTrainingDemo/README.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+
+In this tutorial, we will be fine-tuning a CIFAR 10 model on an android app using ExecuTorch.
+
+## Environment Setup
+
+### Android Environment Setup
+
+Ensure that your development environment meets the following requirements:
+
+1. Minimum CMake version: `CMake 3.22` or higher
```
Comment (jathu) [inline cifar/android/CifarETTrainingDemo/README.md:11]: nit: not needed, ET's install step will take care of it

**Review by jathu — APPROVED** [review]
